### PR TITLE
Use "regular" cart instead SignupCart

### PR DESCRIPTION
**Pull request by yurynix**
**TL;DR:** In this PR I propose replacing `SignupCart` with the "regular" cart.

We've encountered a problem on #11365,
the user couldn't upgrade her domain only site to a site with plan. ( [GIF](https://cloud.githubusercontent.com/assets/699132/22880195/97e6e74c-f1e1-11e6-8582-85c8ffc15924.gif) by @gziolo )

This was caused by the following:
- [SignupCart.addToCart()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/step-actions.js#L266) called with a plan product
- the plan product is _added_ on the server's cart
- user forwarded to checkout
- checkout checks the cart (the "regular" one), but damn! `CartSynchronizer` haven't caught up with the server yet, so it has the old cat version, without the product :sob:
- user redirect back 😞 

This was addressed by @drewblaisdell in [this commit,](https://github.com/Automattic/wp-calypso/pull/11365/commits/634e324443dec959cece489bbd02e3891a94895c) which disables the "regular" cart while we're doing the signup flow, which makes the cart uninitialised in this case, then, when we reach
checkout, checkout requests data from the CartStore, which causes it to initialize.

IMO, as long as we have `CartStore` and `CartSyncrhonizer` used in the codebase, we should treat it as the source of truth for the cart, as long as we bypass it when we're updating it can't do it's job - therefor I think we should drop `SignupCart` and only use `CartStore`.
  
#### Testing instructions
  
1. Run `git checkout add/refactor-signup-cart` and start your server, or open a [live branch](https://calypso.live/?branch=add/refactor-signup-cart)
2. Open the [`Home` page](http://calypso.localhost:3000/)
3. Validate all signup flows works
4. Validate this PR together with #11365
  
#### Additional notes
 
The commit  [e795448](https://github.com/Automattic/wp-calypso/pull/11401/commits/e795448287280bd73e6b85a77f704968787df85b) is not entirely related, this functionality is broken in `master` as well, this test wasn't actually checking what it claims to check.
  
#### Reviews
  
- [ ] Code
- [ ] Product
- [ ] Tests

### CODE REVIEW

File: client/lib/signup/flow-controller.js [hunk 1]
```diff
@@ -130,6 +130,45 @@ assign( SignupFlowController.prototype, {
...
+	 */
+	_assertStepDependencies( step ) {
+		const dependencies = steps[ step.stepName ].dependencies || [];
+		const dependenciesFound = pick( SignupDependencyStore.get(), dependencies );
+		const dependenciesSatisfied = dependencies.length === keys( dependenciesFound ).length;
+		const signupProgress = filter(
+			SignupProgressStore.get(),
+			_step => ( -1 !== this._flow.steps.indexOf( _step.stepName ) )
+		);
+
```
Comment (drewblaisdell) [inline client/lib/signup/flow-controller.js]: I think a lot of this logic is shared with `_canProcessStep`. It might be worth removing it there.

File: client/lib/signup/flow-controller.js [hunk 2]
```diff
@@ -130,6 +130,45 @@ assign( SignupFlowController.prototype, {
 		}.bind( this ) );
 	},
 
+	/**
+	 * This assertion helper validates that there are dependencies for the step provided
+	 * by completed steps
+	 * @param step The step to validate
+	 */
+	_assertStepDependencies( step ) {
```
Comment (drewblaisdell) [inline client/lib/signup/flow-controller.js]: I'm trying to understand why this function is needed, because we currently assert that all dependencies listed as provided in the config are provided in `_assertFlowProvidedRequiredDependencies`. Also, what do you think about using a more specific name here (`_assertStepProvidesDependencies`?).

File: client/lib/signup/flow-controller.js [hunk 3]
```diff
@@ -205,11 +246,12 @@ assign( SignupFlowController.prototype, {
...
 			const apiFunction = steps[ step.stepName ].apiRequestFunction.bind( this );
-
-			apiFunction( ( errors, providedDependencies ) => {
+			const apiFunctionDoneCallback = ( errors, providedDependencies ) => {
 				this._processingSteps[ step.stepName ] = false;
 				SignupActions.processedSignupStep( step, errors, providedDependencies );
-			}, dependenciesFound, step );
+			};
+
+			defer( apiFunction, apiFunctionDoneCallback, dependenciesFound, step );
```
Comment (drewblaisdell) [inline client/lib/signup/flow-controller.js]: Do we need to defer this?

File: client/lib/signup/flow-controller.js (hunk 2 above)
Comment (yurynix) [inline client/lib/signup/flow-controller.js]: Take a look at that test:
https://github.com/Automattic/wp-calypso/pull/11401/files#diff-c530d5162d705f5bd3511598dbc47af2R108

In this scenario we're getting stuck and no error is thrown / displayed, because we haven't completed the flow, 
so we're not getting to the _assertFlowProvidedRequiredDependencies [check](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/flow-controller.js#L168) because the `siteCreation` step is not completed and is waiting for the token that never supplied from the next step.

File: client/lib/signup/flow-controller.js (hunk 3 above)
Comment (yurynix) [inline client/lib/signup/flow-controller.js]: Yes, if we want the apiFunction to be able to dispatch action, since we are in a call graph that started in a dispatch handler.

File: client/lib/signup/flow-controller.js (hunk 1 above)
Comment (yurynix) [inline client/lib/signup/flow-controller.js]: I debated with myself about it, both this method and `_canProcessStep` require those.

Another solution could be remove all of this here and pass them from `_canProcessStep` into this method, but that seems like a lot of parameters to me, so I chose having only the `step` as explicit parameter and deduce everything else, this have the advantage of being more self-contained method

**Comment by Tug** [comment]
> I'm not sure adding complexity to CartSynchronizer and CartStore (which ultimately need to be decoupled) to make them work in signup (which doesn't need to poll the shopping cart) is a good intermediary step.

I haven't spent much time reviewing this PR yet but I must agree on the general idea here. The `CartSynchronizer` is something we would like to move away from, and rather than refactoring it to fit it to our need we should think about porting it to the redux store for instance. Granted this is a much bigger task but we're not fixing any bug here as far as I know so any change on it will simply add more complexity and is a risk to introduce new bugs.

**Comment by gziolo** [comment]
This code is already quite complicated without any new changes introduced :) I can understand why @drewblaisdell and @Tug are proposing to focus our efforts on moving all that code to Redux. Are you sure those changes won't break anything else? I have a feeling like whenever we touch checkout code we fix something in one place and break something else in a completely unrelated place.

**Comment by yurynix** [comment]
@Tug @drewblaisdell and @gziolo Thank you for taking the time to looking at that PR!

First of all, I'd like to highlight that this PR is removes more code than it adds, so it _should_ reduce total LoC ( which is not the same is complexity, but that's really hard to measure objectively IMO )

This PR: `+145 −98`
Commit [f1abfee](https://github.com/Automattic/wp-calypso/pull/11401/commits/f1abfee186f04ea8b8a35491f656ec3dda1a5763): `+56 -4` ( it is not entirely related and can be dropped to a separate PR )
Total _without_ the test commit: **+89 -94**

My second argument is that the current state in `master` is that we are required to [call](https://github.com/Automattic/wp-calypso/pull/11401/files#diff-7b0ca7ad49dc313c6af0636bf22edd4eL130) `disableCart()` everywhere we would like to use "SignupCart" rather than "regular" cart.

the issues with that IMO:
1. We have 2 "things" that handle cart stuff, SignupCart and CartStore, IMO there should be only one.
2. We have a dependency between CartStore and SignupCart in the form of a requirement to remember calling `disableCart()` before the first call to `SignupCart.addToCart()`, that dep. is not that straightforward even for someone who's familiar with the codebase.

While I agree that refactoring CartStore and SignupCart to `redux` is a much better approach, this is a huge amount of work we don't have a roadmap for, yet.

In this PR I propose to meanwhile replace SignupCart (and disableCart [commit](https://github.com/Automattic/wp-calypso/pull/11365/commits/634e324443dec959cece489bbd02e3891a94895c)), with the regular cart that is referenced everywhere else and is essentially the "source of truth" for the cart.

**Comment by yurynix** [comment]
@drewblaisdell 
> Also, do we need to update flow-controller here? It'd probably be best to not update two critical and separate parts of Calypso (lib/signup/flow-controller and lib/cart) unless we need to.

We can drop the [commit](https://github.com/Automattic/wp-calypso/pull/11401/commits/f1abfee186f04ea8b8a35491f656ec3dda1a5763) that updates the assertion, but we'll still have the [defferation](https://github.com/Automattic/wp-calypso/pull/11401/files#diff-9b8826c512c150c2396692c766af5c35R254) change.

@Tug 
> Granted this is a much bigger task but we're not fixing any bug here as far as I know so any change on it will simply add more complexity and is a risk to introduce new bugs.

This PR is a suggestion for replacing a FIX to a bug that was introduced in #11365 with this [commit](https://github.com/Automattic/wp-calypso/pull/11365/commits/634e324443dec959cece489bbd02e3891a94895c).
I've discussed that a bit with @drewblaisdell on slack and we decided to push #11365 forward and discuss alternatives to that in a separate PR, so here it is =)

@gziolo 
> Are you sure those changes won't break anything else? I have a feeling like whenever we touch checkout code to fix something we break something else in a completely unrelated place.
How one can be sure 100% ? We should review it and test it with different flows and add tests for the still uncovered parts. 
However the only major change in `CartSynchronizer` is an addition of calling a callback when the cart is flushed to the server, which is used only in the signup flows.

File: client/lib/cart/store/cart-synchronizer.js [hunk 4]
```diff
@@ -208,8 +227,16 @@ CartSynchronizer.prototype._performRequest = function( type, requestFunction ) {
...
 
 		if ( error ) {
+			if ( request.type === 'update' ) {
+				this._processServerFlushCallbacks( error, null );
+			}
 			throw error;
 		}
+
+		if ( request.type === 'update' ) {
+			this._processServerFlushCallbacks( null, error );
```
Comment (Tug) [inline client/lib/cart/store/cart-synchronizer.js]: `this._processServerFlushCallbacks( null, newValue );` here I believe

File: client/lib/cart/store/cart-synchronizer.js [hunk 5]
```diff
@@ -106,17 +107,27 @@ CartSynchronizer.prototype.update = function( changeFunction ) {
...
+};
+
+CartSynchronizer.prototype._processChangeFunction = function ( changeFunction ) {
 	if ( this._activeRequest && this._activeRequest.state !== 'canceled' ) {
 		this._activeRequest.state = 'canceled';
 	}
 
 	this._latestValue = changeFunction( this._latestValue );
 	this._performRequest( 'update', this._postToServer.bind( this ) );
-	this.emit( 'change' );
```
Comment (Tug) [inline client/lib/cart/store/cart-synchronizer.js]: Why not emit events instead of queuing callbacks?

Comment (Tug) [inline client/lib/cart/store/cart-synchronizer.js]: I would call that event `synced` rather than `flush` but maybe that's a personal preference

File: client/signup/main.jsx [hunk 6]
```diff
@@ -124,11 +123,6 @@ const Signup = React.createClass( {
 		} );
 		recordSignupStart();
 
-		// Signup updates the cart through `SignupCart`. To prevent
-		// synchronization issues and unnecessary polling, the cart is disabled
-		// here.
-		disableCart();
```
Comment (Tug) [inline client/signup/main.jsx]: We still need to empty the cart here don't we?
In case we'll be reusing stale data from another signup or if a flow uses an existing site.

File: client/lib/cart/store/cart-synchronizer.js (hunk 5 above)
Comment (yurynix) [inline client/lib/cart/store/cart-synchronizer.js]: This is git playing games, `this.emit( 'change' )` is at the same place it was before, at the end of `update()` function.

> Why not emit events instead of queuing callbacks?

currently we call it threw an actionCreator, like so:
```js
addItems( [ domainItem ], 'signup', error => callback( error, providedDependencies ) );
```

So i'm not sure how emitting events will work, could you maybe show an example?

File: client/lib/cart/store/cart-synchronizer.js (hunk 4 above)
Comment (yurynix) [inline client/lib/cart/store/cart-synchronizer.js]: Amazing catch, thank you!

File: client/lib/cart/store/cart-synchronizer.js (hunk 5 above)
Comment (Tug) [inline client/lib/cart/store/cart-synchronizer.js]: I guess it's fine as it is, my idea was to emit `synced` event in the CartSynchronizer that the CartStore could listen to.
So replace this call:
```js
_synchronizer.update( wrappedFunction, serverFlushCallback );
```
by:
```js
_synchronizer.update( wrappedFunction );
_synchronizer.once( 'synced', serverFlushCallback );
```

Comment (yurynix) [inline client/lib/cart/store/cart-synchronizer.js]: That is something we can do in addition, but we still need the callback thing to integrate with the signup framework IMO, to hook it [here](https://github.com/Automattic/wp-calypso/pull/11401/files/0d0fcb1dbbfe1363a0ba97c3fdb4e7b2c85e444d#diff-160d380c201b4db2d947c495e877a98aR44), I'm not sure how would we do it with the even you're suggesting.

File: client/signup/main.jsx (hunk 6 above)
Comment (yurynix) [inline client/signup/main.jsx]: Thats a complicated question really, I'm not sure what's the answer to this one.

From a certain point of view you _are_ right, we should clear the cart, from another point of view no, we should do something else if user has items in his cart, like, maybe, ask him if he want to check-out using this items?

I'm not sure what's the answer here.

BTW, that is also true for the previous approach with `disableCart()`, disabling the cart isn't clearing the cart, just disabling the fetch polling, but if the cart is full from a prev. "stale data" as far as I understand.

File: client/lib/cart/store/cart-synchronizer.js (hunk 5 above)
Comment (yurynix) [inline client/lib/cart/store/cart-synchronizer.js]: Also, maybe I'm missing something, but how can I be sure that the emitted 'synced' event is corresponding with the same `wrappedFunction` I called `update()` with ? As far as I understand that is not guaranteed, because if a request is in progress, the `wrappedFunction` of `update()` will be enqueued and `synced` event will be emitted for the previous completion sync, then, the enqueued `wrappedFunction` will proceed.

File: client/signup/main.jsx (hunk 6 above)
Comment (aidvu) [inline client/signup/main.jsx]: :+1: That's how I understood it too. Disable just turns off polling until we dispatch for a cart again. Cart is cleared on the back-end if successful. Not sure how to handle stale carts, apart from clearing them every time for a `signup` cart.

File: client/lib/signup/step-actions.js [hunk 7]
```diff
@@ -41,7 +41,7 @@ function createCart( callback, dependencies, data ) {
 			themeItem
 		};
 
-		SignupCart.addToCart( cartKey, [ domainItem ], error => callback( error, providedDependencies ) );
+		addItems( [ domainItem ], 'signup', error => callback( error, providedDependencies ) );
```
Comment (aidvu) [inline client/lib/signup/step-actions.js]: This will break PayPalExpress. Should use `cartKey` here, or figure how to do it in the backend.

File: client/lib/signup/step-actions.js [hunk 8]
```diff
@@ -98,9 +98,7 @@ function createSiteWithCart( callback, dependencies, {
 			].filter( item => item );
 
 			if ( newCartItems.length ) {
-				SignupCart.addToCart( siteId, newCartItems, function( cartError ) {
-					callback( cartError, providedDependencies );
-				} );
+				addItems( newCartItems, 'signup', cartError => callback( cartError, providedDependencies ) );
```
Comment (aidvu) [inline client/lib/signup/step-actions.js]: Not sure I understand why this change and the one bellow?

File: client/lib/signup/step-actions.js (hunk 7 above)
Comment (aidvu) [inline client/lib/signup/step-actions.js]: Ooops, this is context. :man_facepalming:

File: client/lib/signup/step-actions.js (hunk 8 above)
Comment (aidvu) [inline client/lib/signup/step-actions.js]: :sob:

**Comment by aidvu** [comment]
Breaks the regular flow. Now you past the response of the `cart` endpoint as `data`, and [assertValidDependencies](https://github.com/Automattic/wp-calypso/blob/4a0962e58109cd517e9db6b163d4ad4fc5a06f02/client/lib/signup/dependency-store.js#L34) on `SUBMIT_SIGNUP_STEP` will find `extraItems`.

**Comment by yurynix** [comment]
After discussing that some more, we've decided not pursuit it further.
